implement library for PMI simple v1 wire protocol server side

garlick opened the pull request:
This PR implements a small "protocol engine" for the PMI simple v1 wire protocol, for the server side and then uses it and the recently added `subprocess_socketpair()` call to implement PMI as a bootstrap option for brokers in `flux-start`.

We could go a step further and eliminate the `--rank, --size, and --boot-method=LOCAL` code in the broker and make this the default way that `flux-start` operates, but I stopped short of that.

It's worth noting that this subset of PMI1 lacks `PMI_Get_id()` (used to communicate a non-numeric job id) and `PMI_Get_clique_ranks()` (used to determine which ranks can communicate over ipc versus tcp).  

Also, in this iteration I didn't bother implementing the protocol operations for `PMI_Publish_name()` , `PMI_Unpublish_name()`, `PMI_Lookup_name()`, or `PMI_Spawn_multiple()`.  (These are also stubbed out in libpmi-client)

A demo of `flux-start --boot-method=PMI` was added to the sharness hydra test.

Another test was added that exercises each protocol operation, pitting libpmi-client against libpmi-server over a single socketpair.

grondo commented:
This looks pretty reasonable to me! I'd be fine with removing broker `--rank` `--size` etc options if they are no longer needed!

So does this mean we're close to running MPICH jobs with PMI simple? (Is there some work for wrexecd (and thus its replacement) to support this?)

garlick commented:
OK, maybe I'll go ahead and remove that code and make PMI the default.  Why not?  :-)

On running MPICH the todo list is I think:
* convert wrexecd over to libsubprocess
* add a `kvs_fence()` variant with a continuation interface
* implement reactive PMI service in wrexecd/lua plugin
* implement option and/or heuristic to activate PMI service

grondo commented:
> convert wrexecd over to libsubprocess

I was hoping not to have to worry about that for wreck prototype.. could we hack around that somehow for now?

garlick commented:
I was actually thinking we would wait until the new program service, but it is pretty close.  Maybe we should see if there is some clever way to get it going without changing too much.  (Not feeling super clever right at the moment)

garlick commented:
Just rebased with PMI as the only method used by flux-start, and dropped the explicit test for running that way.  A few options were dropped from the broker, and its man page was updated.  Finally, check a boundary case for `PMI_KVS_Get()` of a large value in the new pmi/simple.t test.

grondo commented:
Great! 

One quick comment. It looks like there are 3 copies of a `readline` function in this PR?  Maybe that is enough to create a function in libcommon?

Or, better yet, we already have cbuf in liblsd, and that has a good buffered readline implementation, so you wouldn't be reading a character at a time. However, I don't know if it is worth using that implementation for this or not...

garlick commented:
Agreed, that's probably the lamest part of this PR and it really should be replaced with reactive, buffered I/O.  I'll have a look tomorrow.

grondo commented:
@garlick -- I was thinking of the cbuf being moved into the `pmi_simple_server` implemtation -- either by feeding data directly into the server or by having the socket fd be part of the server too (then, you just register an fdwatcher internally in `pmi_simple_server` and it manages its own state). I see now, however, that is a pretty  big change, and I didn't even look closely enough to see if it was possible to implement for all cases! Sorry!

I think what is here is great for now, and maybe when we try using this with MPI as part of wrexecd or the replacement, we could look at moving the buffered IO into pmi server implementation, then just wire it directly into wrexecd/shell reactor? (At least that way there won't be duplication of code among users)

garlick commented:
Another thought would be to attach a zio to the socketpair in libsubprocess, and generalize `subprocess_write()`  and `subprocess_io_cb_f` since libsubprocess is already reactor-aware?

grondo commented:
Yeah, I guess that would result in the least code probably, however hopefully it would get too awkward. Also, that would squash any hope of using pmi_simple_server in anyplace that doesn't use libsubprocess.

garlick commented:
Hmm.  Maybe zio by itself could replace the reactor-to-cbuf glue here?

grondo commented:
I have this feeling that the design of zio is a bit awkward. Should we take this opportunity to look at a redesign of the interface?

garlick commented:
I reverted back to the simpler, synchronous I/O implementation as I haven't really gotten anywhere good with this.

The cbuf based implementation was overly complicated and I felt there were a lot of unexplored corners in the code.  The zio based implementation was a little better, but awkward for a bidirectional socket (socket was getting double closed inside zio).  It seems like we do need a generalized abstraction for buffered, reactive I/O which might be a (redesigned?) zio, but perhaps we can  tackle that another day?

garlick commented:
Actually this needs to be rebased on the current master, hang on.

garlick commented:
OK, rebased.

garlick commented:
I merged PR #459, squashed some incremental development state, and rebased again.  If this passes travis-ci I think it's ready for a merge.

grondo commented:
Great!
